### ISSUE TITLE
Possible false positive for L003 in jinja template

### ISSUE THREAD

**Issue by igorcalabria** [sqlfluff/sqlfluff#3750]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

When using if-else blocks like this: 

```
{% if true %}
    {% set something = 'something' %}
{% else %}
    {% set something = 'something else' %}
{% endif %}
```

I'm getting L003 Violations
```
L:   2 | P:   5 | L003 | Expected 0 indentations, found 1 [compared to line 01]
```

### Expected Behaviour

I expected the linter to accept indented control blocks and not trigger a violation.

### Observed Behaviour

sqlfluff triggered a L003 violation for an indented control block

### How to reproduce

`test.sql`
```sql
{% if true %}
    {% set something = 'something' %}
{% else %}
    {% set something = 'something else' %}
{% endif %}


SELECT {{ something }} AS thing
```
sqlfluff lint --dialect ansi test.sql

### Dialect

ansi

### Version

sqlfluff, version 1.2.1

### Configuration

I'm using default configuration

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by pdebelak** [comment]
Of note, the following doesn't have lint errors:

```sql
{% if true -%}
    {% set something = 'something' %}
{% else -%}
    {% set something = 'something else' %}
{% endif %}


SELECT {{ something }} AS thing
```

**Comment by igorcalabria** [comment]
That makes sense. It's stripping the leading whitespace(the new line). Jinja has a configuration that may enable this automatically: https://jinja.palletsprojects.com/en/3.1.x/templates/#whitespace-control  `trim_blocks` and `lstrip_blocks`. Does it make sense to expose this configuration to the end user? I can't speak for everyone but our models are pretty light on macros and I believe that stripping by default is a saner option to us. To keep things backwards compatible, sqlfluff just need to set the defaults to False

**Comment by barrywhart** [comment]
Trimming leading whitespace won't solve this issue in general. This file reports a similar error, but on line 4.

```
SELECT 1;

{% if true %}
    {% set something = 'something' %}
{% else %}
    {% set something = 'something else' %}
{% endif %}

SELECT 2;
```

Perhaps L003 can be updated to behave appropriately for "template-only" lines in these situations.

**Comment by pdebelak** [comment]
Yeah, I agree that trimming leading whitespace _shouldn't_ make any difference here to rule L003 since either way the jinja block has no effect on the outputted sql and thus the whitespace trimming is irrelevant. In general, does/should L003 care about the indentation of the rendered sql (personally, I don't really care about that, but I know some people do) or the indentation of the jinja (something I do care about, but might be sort of out of scope for sqlfluff)?

**Comment by igorcalabria** [comment]
@pdebelak 
> the jinja block has no effect on the outputted sql

Being a little pedantic, I don't think that's 100% accurate. It has no effect on the sql tree, but it does have an effect com on the rendered sql. For example: 

```
SELECT 1;

{% if true %}
    {% set something = 'something' %}
{% else %}
    {% set something = 'something else' %}
{% endif %}

SELECT 2;
```

gets rendered (using default jinja options in a stand-alone script, not in the context of sqlfluff) as 

```
SELECT<space>1;<newline>
<newline>
<newline>
<space><space><space><space><newline>
<newline>
<newline>
SELECT<space>2;<newline>
```
If sqlfluff is testing L003 against the rendered code, it almost makes sense to report a L003 violation on line 4 because it looks like a indentation even if it's only made of whitespace. I do like your ideia of simply ignoring the rendered code as much as possible.

**Comment by alanmcruickshank** [comment]
I'll take a look at this one - because I think it's related to some of the work that I'm already doing.

**Comment by alanmcruickshank** [comment]
I think this is due to mis-parsing of the `{% set %}` command or a bug in the interpretation of the indents implied by the `{% if %}` tag. My gut says the former. Here's the parse output for the file (on the most recent release):

```
sqlfluff [main ≡]> sqlfluff parse test.sql -d ansi
[L:  1, P:  1]      |file:
[L:  1, P:  1]      |    [META] placeholder:                                       [Type: 'block_start', Raw: '{% if true %}']
[L:  1, P: 14]      |    newline:                                                  '\n'
[L:  2, P:  1]      |    whitespace:                                               '    '
[L:  2, P:  5]      |    [META] placeholder:                                       [Type: 'block_start', Raw: "{% set something = 'something' %}"]
[L:  2, P: 38]      |    newline:                                                  '\n'
[L:  3, P:  1]      |    [META] placeholder:                                       [Type: 'block_mid', Raw: '{% else %}']
[L:  4, P:  5]      |    [META] placeholder:                                       [Type: 'block_start', Raw: "{% set something = 'something else' %}"]
[L:  5, P:  1]      |    [META] placeholder:                                       [Type: 'block_end', Raw: '{% endif %}']
[L:  5, P: 12]      |    [META] end_of_file:
```

I don't think the `{% set %}` statements should have the type `block_start` - but we don't really have a provision for a different type at the moment in the code.

**Comment by barrywhart** [comment]
There are two "flavors" of `{% set %}`:
* "One line": `{% set myvar = "myvalue" %}`
* Block style:
```
{% set myvar %}
myvalue
{% endset %}
```

Probably we need to distinguish the two when setting types. (We already distinguish it for other reasons, so this won't be _completely_ new work.)

**Comment by lyschoening** [comment]
We are also experiencing this issue. `{% set x = ... %}` is a common pattern in dbt, so this breaks indentation rules in a lot of our files. We usually don't have the option to use the block style notation as we are setting values other than strings. 

The issue is amplified by the evaluation of the rule being inconsistent. For example, `{% set x = [] %}` and `{% set x = [1] %}` evaluate differently.
